### DISCUSSION TITLE
End of charging not working/keeps jumping between charge/discharge

### DISCUSSION THREAD

**Discussion by joseal** [Louisvdw/dbus-serialbattery#732] [Q&A] [answered]
Hi,
I'm using this driver with the following setup:

Multiplus II in AC coupling + ESS 
JKBMS connected via RS485
16 x 304ah cells (48v)

And in general, it has been working fine (it controls charge/discharge current, etc...), but for some reason it never ends the charging cycle, as it stays jumping between charge/discharge, with Mutliplus II in absorption and driver reporting bulk...

Example video:
https://github.com/Louisvdw/dbus-serialbattery/assets/24917156/d1650161-2a21-46fd-a0d7-5badbb347cf6

I believe it should be a setup issue, but I'm not finding it...

Please let me know what other info you need to help me!

**Comment by mr-manuel** [comment]
Please check directly the remote console/GUI and the cell voltages as also the parameters page.

**Reply by joseal** [reply]
Hi Manuel,

in first place, thank you very much for your reply... I really need some help, because I'm running out of ideas.
Currently, the battery is not charged, so is not showing the right values, but later today I will post those screens. 

In the meantime, please clarify if my understanding is right...

When the baterry is full (16x3.45v = 55.2v), my expectation is to see the driver switch from bulk to absorption (as well the multiplus lights) and after 1h or so, it should go idle, as no more charge is needed.

However, what I've is: Multiplus changing to absorption around the 55.2v, the driver stays in bulk mode and with that strange behavior, in a loop jumping between charging and discharging with ~100w, until the sun goes down and the real discharge cycle starts (during thr summar is in this state for ~10h, what conserns me, that it could be causing a lot of stress to the cells).

It seems like a conflict/dispute between multi and the driver, trying to take control...

I've forced the battery monitor to the driver, instead of automatic... I've removed battery monitor flag from VE.config... but no changes...

As an example, it stayed 7 hours:

![image](https://github.com/Louisvdw/dbus-serialbattery/assets/24917156/1c72f1a1-35e4-4de0-a06c-2dc03364fce7)

Later today, I will post my config...

**Reply by mr-manuel** [reply]
Search a bit in the issues and discussions. The discharge/charge behaviour is pretty normal since it's needed for system stability from Victron. The MultiPlus should show external control.
Please read the config.default.ini file. There are a lot of useful informations.

**Comment by joseal** [comment]
Hi Manuel,

today, I did some progress...

basically, by making both battery measurements visible:

![image](https://github.com/Louisvdw/dbus-serialbattery/assets/24917156/315c52fe-2460-4997-b8a8-0699117b2e35)

I noticed that there was a 0.13v deviation, between the measurements done by the BMS and the Multiplus. 
I've checked with a percision multimeter on both sides (at the exit of the multi and at the battery terminals) and the BMS was well calibrated (the deviation is at the Multiplus).

This explains why the battery was never reaching the 55,2v, as the multi was being told to supply 55,2v, but in reality, it was only providing 55.08v (leaving the driver stuck in bulk).

![image](https://github.com/Louisvdw/dbus-serialbattery/assets/24917156/5ea1d1f8-784f-4b00-a2e9-e1416fd34e98)

So, reading about it, I found that enabling the shared voltage sense, I was able to have the multi using the value aligned with the BMS value:

![image](https://github.com/Louisvdw/dbus-serialbattery/assets/24917156/3162ead2-fc21-4466-93e9-1edd6c6b939b)

![image](https://github.com/Louisvdw/dbus-serialbattery/assets/24917156/d45dbc0e-0f86-4ee4-a7c7-e4c5e4f9ebbe)

After this change, today I was finally able to see the driver switch to Float mode!

In the meantime, the sun went way and I was not able to finish the tests, but tomorrow I will continue....

However, I've already a couple of doubts:

- I've this option selected in VE-config... considering that I've JK-BMS connected with ModBus, should it be different?

![VE-config-5](https://github.com/Louisvdw/dbus-serialbattery/assets/24917156/02d8db14-0c6c-4f98-a7ba-373fb0ea8f60)

- Should I've this ESS mode or should it be in external control?

![image](https://github.com/Louisvdw/dbus-serialbattery/assets/24917156/914a2bd7-edbd-4d20-9915-9bdef6b53158)

- Do you know when/how JK-BMS resets the SOC? I'm not sure if mine is right...

**Reply by mr-manuel** [reply]
1. Thats correct
2. External control not what you think. This is when you want control the ESS by yourself via external software. Select one of the top three, if you don't know what this excactly does.
3. See https://github.com/Louisvdw/dbus-serialbattery/issues/703

**Comment by joseal** [comment]
Hi Manuel,

once more, thank you very much for your reply!

After enabled the Shared Voltage Sense setting, the driver is not working just fine!
I can see it moving from Bulk to Absorption (for a very little time) and then goin on Float:

![image](https://github.com/Louisvdw/dbus-serialbattery/assets/24917156/d76890de-8856-4c32-a9d6-dad43ddd3ae2)

One quick question: 
Why charge limitation, shows only "Cell Voltage" ? Shouldn't be similar to discharge limitation? 

root@einstein:~# tail -F /data/log/dbus-serialbattery.ttyACM0/current
@400000006497119520a794bc INFO:SerialBattery:> MAX BATTERY CHARGE CURRENT: 70.0A | MAX BATTERY DISCHARGE CURRENT: 100.0A (read from BMS)
@400000006497119520a7a45c INFO:SerialBattery:> CVCM:     True
@400000006497119520a7b014 INFO:SerialBattery:> MIN CELL VOLTAGE: 2.9V | MAX CELL VOLTAGE: 3.45V
@400000006497119520a7bbcc INFO:SerialBattery:> CCCM CV:  True  | DCCM CV:  True
@400000006497119520a7c784 INFO:SerialBattery:> CCCM T:   True  | DCCM T:   True
@400000006497119520aa1d2c INFO:SerialBattery:> CCCM SOC: True  | DCCM SOC: True
@400000006497119520aa30b4 INFO:SerialBattery:Serial Number/Unique Identifier: JK_B2A20S20P
@400000006497119524a47094 INFO:SerialBattery:DeviceInstance = 1
@400000006497119524b237ec INFO:SerialBattery:com.victronenergy.battery.ttyACM0
@4000000064971195267d7e24 INFO:SerialBattery:publish config values = 1

**Reply by mr-manuel** [reply]
(Dis)Charge Limitation shows the limiting component. In your case CCL is only limited by high cell voltage. DCL is limited by config limit, cell voltage, temp and SoC equally. This because you set on your BMS a discharge limit of 100A and in the config 80A.

**Comment by joseal** [comment]
Hi Manuel,
sorry, but I'm not understanding this logic...

Now, I've changed the BMS setting: 

![image](https://github.com/Louisvdw/dbus-serialbattery/assets/24917156/46c0547b-0048-44b7-8649-929a09261b88)

And I've this:

![image](https://github.com/Louisvdw/dbus-serialbattery/assets/24917156/349e3409-fb8e-4b04-a20a-74e248215a3b)

**Reply by mr-manuel** [reply]
Where is the problem?

**Comment by joseal** [comment]
Hi Manuel,

sorry for being insisting, but maybe I'm reading wrong the meaning of the "charge limitation" and discharge limitation", so please correct me:

current battery status:

  Voltage = 55.21v
  SOC = 82% (completly messed up, but I'm need to test the steps you mentioned in #703, to see if I can fix it)
  Temp = 26ºC

my config.ini:

```
MAX_BATTERY_CHARGE_CURRENT = 60.0
MAX_BATTERY_DISCHARGE_CURRENT = 80.0
``` 

current BMS limits:
Charge: 70a
Discharge: 100a

And I've this:

![image](https://github.com/Louisvdw/dbus-serialbattery/assets/24917156/8f1dcd55-5397-4d39-befa-7ed014a12f3c)

So, we've "charge limitation" = "cell voltage", because we are niether getting limited by SOC (as it <91%) or Temp (as it is between 25-35ºC). So, only the voltage is limiting. This makes sense to me!

But, why we've "discharge limitation" = "Config limit, Cell Voltage, Temp & SoC"? Config limit makes sense, as I've my BMS setup for 100a and the driver is limiting it to 80a, but for example Temp is 26ºC (we are between the 15-45ºC, where fraction should be = 1). Voltage should be also ok (=>3.45) and SoC >30%...

SoC is actually causing a problem, because as it is wrong, it is blocking floating... but this is not an issue in the driver logic! 

![image](https://github.com/Louisvdw/dbus-serialbattery/assets/24917156/90167a8c-62f4-453d-b38d-7d8a2d7cb429)

If I match the BMS discharge limit with the value in the config.ini (MAX_BATTERY_DISCHARGE_CURRENT = 80.0), it seems to point to the currect limitation:

![image](https://github.com/Louisvdw/dbus-serialbattery/assets/24917156/887119fe-6dd0-453a-b7e4-19625d06819c)

Again, sorry if I'm reading it completly wrong, but this is why I'am confused.

Once, more thank you very much for your patiance and great work!

**Reply by mr-manuel** [reply, marked as answer]
Since you have enabled all options and all options are calculated as fractions from the MAX_DISCHARGE_LIMIT all options are limiting the discharge. If we write unlimited it would be the wrong description, but this is correct as it is.

https://github.com/Louisvdw/dbus-serialbattery/blob/86a6b956e15bee21d5a298283c9d121f0eb82a02/etc/dbus-serialbattery/config.default.ini#L86-L152

**Reply by joseal** [reply]
Hi Manuel,
I think we can close this thread, as the issue has been identified "the voltage difference between the Multi and the BMS"
Regarding the "Discharge Limitation" behavior, once I've a few more time (work is currently not allowing me to do that much), I need to look at logic you have in the source code, to see if I understand how it better...
Charge Limitation seems to be working fine, as I can see it changing between "Cell Voltage", "SoC", etc...
Once more, thank you very much for your support and great work!
